### PR TITLE
Promise storage

### LINKED ISSUE #715: Promise storage
We need to implement a storage that allows for storing issued promises. Will probably end up storing these on bolt. The storage must be thread-safe and provide atomic increases in promise sequenceIDs for a given issuer.

### PR DESCRIPTION
**Pull request by vkuznecovas**
Promises are now stored on boltdb. Touched a little on balance tracker as well.

CLOSES #715

### CODE REVIEW

File: session/balance/provider/balance_tracker.go [hunk 1]
```diff
@@ -66,12 +63,18 @@ func (bt *BalanceTracker) calculateBalance() {
...
+}
+
+// Start starts keeping track of time for balance
+func (bt *BalanceTracker) Start() {
+	bt.timeKeeper.StartTracking()
+}
+
+// Add increases the current balance by the given amount
+func (bt *BalanceTracker) Add(amount uint64) {
+	bt.totalPromised += amount
```
Comment (soffokl) [inline session/balance/provider/balance_tracker.go]: Is there any possibility to have a race condition here for `totalPromised`? Maybe we need to protect it somehow?

File: session/payment/session_balance.go [hunk 2]
```diff
@@ -97,9 +124,88 @@ func (ppo *SessionBalance) Start() error {
 	}
 }
 
+func (ppo *SessionBalance) loadInitialPromiseState() (promise.StoredPromise, error) {
```
Comment (soffokl) [inline session/payment/session_balance.go]: Why does it call `ppo` for `*SessionBalance`?

File: session/payment/session_balance.go [hunk 3]
```diff
@@ -97,9 +124,88 @@ func (ppo *SessionBalance) Start() error {
...
+	lastPromise, err := ppo.promiseStorage.GetLastPromise(ppo.issuer)
+	if err != nil {
```
Comment (soffokl) [inline session/payment/session_balance.go]: We can do it a bit smaller:
```golang
	if err != nil {
		// if an error occurs when fetching the last promise, issue a new id
		ppo.sequenceID, err = ppo.promiseStorage.GetNewSeqIDForIssuer(ppo.issuer)
		return lastPromise, err
	} 
	ppo.sequenceID = lastPromise.SequenceID
	return lastPromise, nil
```

File: session/payment/session_balance.go [hunk 4]
```diff
@@ -97,9 +124,88 @@ func (ppo *SessionBalance) Start() error {
...
+	ppo.balanceTracker.Add(amountToAdd)
+	ppo.balanceTracker.Start()
+}
+
+func (ppo *SessionBalance) calculateUnconsumedPromiseAmount(balance uint64, lastPromise promise.StoredPromise) (uint64, error) {
+	// if we're ever in a situation where the unconsumed amount is zero, but the balance is not - something is definitely not right
+	if lastPromise.UnconsumedAmount == 0 && balance != 0 {
+		return 0, fmt.Errorf("unconsumed amount is 0, while balance is %v", balance)
+	}
+	unconsumed := lastPromise.UnconsumedAmount - (lastPromise.UnconsumedAmount - balance)
```
Comment (soffokl) [inline session/payment/session_balance.go]: It's equal to:
```golang
	unconsumed := balance
```
Probably something wrong here.

File: session/payment/session_balance.go [hunk 5]
```diff
@@ -57,6 +68,11 @@ type SessionBalance struct {
 	chargePeriod       time.Duration
 	promiseWaitTimeout time.Duration
 	promiseValidator   PromiseValidator
+	promiseStorage     PromiseStorage
+	issuer             identity.Identity
+
+	sequenceID  uint64
+	lastBalance uint64
```
Comment (soffokl) [inline session/payment/session_balance.go]: `lastBalance` is unused.

File: session/payment/session_balance_test.go [hunk 6]
```diff
@@ -18,61 +18,44 @@
...
 		promiseChannel,
 		time.Millisecond*1,
 		time.Millisecond*1,
 		mpv,
+		mps,
+		issuer,
 	)
 }
 
 func Test_ProviderPaymentOchestratorStartStop(t *testing.T) {
```
Comment (soffokl) [inline session/payment/session_balance_test.go]:
```suggestion
func Test_ProviderPaymentOrchestratorStartStop(t *testing.T) {
```

File: session/payment/session_balance_test.go [hunk 7]
```diff
@@ -81,38 +64,43 @@ func Test_ProviderPaymentOchestratorStartStop(t *testing.T) {
 }
 
 func Test_ProviderPaymentOchestratorSendsBalance(t *testing.T) {
```
Comment (soffokl) [inline session/payment/session_balance_test.go]:
```suggestion
func Test_ProviderPaymentOrchestratorSendsBalance(t *testing.T) {
```

File: session/payment/session_balance_test.go [hunk 8]
```diff
@@ -81,38 +64,43 @@ func Test_ProviderPaymentOchestratorStartStop(t *testing.T) {
...
-	orch := NewMockSessionBalance(MPV)
+	orch := NewMockSessionBalance(MPV, MPS, MBT)
 	defer orch.Stop()
 	go orch.Start()
 
 	assert.Exactly(t, balance.Message{SequenceID: 1, Balance: 0}, <-BalanceSender.balanceMessages)
 }
 
 func Test_ProviderPaymentOchestratorSendsBalance_Timeouts(t *testing.T) {
```
Comment (soffokl) [inline session/payment/session_balance_test.go]:
```suggestion
func Test_ProviderPaymentOrchestratorSendsBalance_Timeouts(t *testing.T) {
```

File: session/payment/session_balance_test.go [hunk 9]
```diff
@@ -81,38 +64,43 @@ func Test_ProviderPaymentOchestratorStartStop(t *testing.T) {
...
 		assert.Equal(t, ErrPromiseWaitTimeout, err)
+		testDone <- struct{}{}
 	}()
 
 	//consume message but never respond
 	<-BalanceSender.balanceMessages
+	<-testDone
 }
 
 func Test_ProviderPaymentOchestratorInvalidPromise(t *testing.T) {
```
Comment (soffokl) [inline session/payment/session_balance_test.go]:
```suggestion
func Test_ProviderPaymentOrchestratorInvalidPromise(t *testing.T) {
```

File: session/promise/storage.go [hunk 10]
```diff
@@ -0,0 +1,181 @@
...
+
+	"github.com/mysteriumnetwork/node/identity"
+)
+
+const promiseBucketPrefix = "stored-promise-"
+const firstPromiseID = uint64(1)
+
+var (
+	// ErrPromiseNotFound represents the error we return when trying to update a non existing promise
+	ErrPromiseNotFound = errors.New("Promise not found")
```
Comment (soffokl) [inline session/promise/storage.go]:
```suggestion
	ErrPromiseNotFound = errors.New("promise not found")
```

File: session/promise/storage.go [hunk 11]
```diff
@@ -0,0 +1,181 @@
...
+}
+
+// NewStorage returns a new instance of promise storage
+func NewStorage(storage Storer) *Storage {
+	return &Storage{
+		storage: storage,
+	}
+}
+
+// StoredPromise is a representation of a promise in storage. It stores the message that the consumer sent,
```
Comment (soffokl) [inline session/promise/storage.go]:
```suggestion
// StoredPromise is a representation of a promise in storage. It stores the message that the consumer sent.
```

File: session/promise/storage_test.go [hunk 12]
```diff
@@ -0,0 +1,411 @@
...
+	"time"
+
+	"github.com/mysteriumnetwork/node/identity"
+	"github.com/stretchr/testify/assert"
+)
+
+var (
+	id       = identity.FromAddress("0x0")
+	timeMock = time.Now()
+	mock     = map[string][]StoredPromise{
```
Comment (soffokl) [inline session/promise/storage_test.go]:
```suggestion
	mock     = map[string][]StoredPromise{
		getBucketNameFromIssuer(id): {{
			SequenceID: 1,
			AddedAt:    timeMock,
			UpdatedAt:  timeMock,
		}},
	}
```

File: session/payment/session_balance.go (hunk 4 above)
Comment (vkuznecovas) [reply session/payment/session_balance.go]: I actually LOL'ed. Leftover from extensive rework of this part, will fix.

File: session/balance/provider/balance_tracker.go (hunk 1 above)
Comment (vkuznecovas) [reply session/balance/provider/balance_tracker.go]: At the moment, I'm not at all concerned by race conditions here. But i'll add a lock to make it atomic anyway.

File: session/balance/provider/balance_tracker_test.go [hunk 13]
```diff
@@ -0,0 +1,77 @@
...
+	mockMoney := money.Money{
+		Amount:   1,
+		Currency: money.CURRENCY_MYST,
+	}
+	mtk := &mockTimeKeeper{elapsed: mockTime}
+	mac := &mockAmountCalculator{toReturn: mockMoney}
+	tracker := NewBalanceTracker(mtk, mac, initialBalance)
+	balance := tracker.GetBalance()
+
+	balanceAfter := initialBalance - mockMoney.Amount
```
Comment (Waldz) [inline session/balance/provider/balance_tracker_test.go]: Static values in asserts liek `1` and `2` would make test more readable, I would say

File: session/payment/session_balance.go [hunk 14]
```diff
@@ -75,48 +92,129 @@ func NewSessionBalance(
...
+func (sb *SessionBalance) startBalanceTracker(lastPromise promise.StoredPromise) {
+	amountToAdd := lastPromise.UnconsumedAmount
+
+	sb.balanceTracker.Add(amountToAdd)
+	sb.balanceTracker.Start()
 }
 
-func (ppo *SessionBalance) receivePromiseOrTimeout() error {
+func (sb *SessionBalance) sendBalance() error {
+	b := sb.balanceTracker.GetBalance()
```
Comment (Waldz) [inline session/payment/session_balance.go]: Unreadable vars looks like `a` and `b`

Comment (vkuznecovas) [reply session/payment/session_balance.go]: renamed

**Pull request merged**